server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,292 @@
+                length
+            ),
+            fileCache
+        );
+    }
+
+    /**
+     * Returns names of all files stored in this directory in sorted order
+     * Does not include locally stored block files (having _block_ in their names)
+     *
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public String[] listAll() throws IOException {
+        logger.trace("listAll() called");
+        String[] localFiles = localDirectory.listAll();
+        logger.trace("Local Directory files : {}", () -> Arrays.toString(localFiles));
+        Set<String> allFiles = new HashSet<>(Arrays.asList(localFiles));
+        String[] remoteFiles = getRemoteFiles();
+        allFiles.addAll(Arrays.asList(remoteFiles));
+        logger.trace("Remote Directory files : {}", () -> Arrays.toString(remoteFiles));
+        Set<String> localLuceneFiles = allFiles.stream()
+            .filter(file -> !FileType.isBlockFile(file))
+            .collect(Collectors.toUnmodifiableSet());
+        String[] files = new String[localLuceneFiles.size()];
+        localLuceneFiles.toArray(files);
+        Arrays.sort(files);
+        logger.trace("listAll() returns : {}", () -> Arrays.toString(files));
+        return files;
+    }
+
+    /**
+     * Removes an existing file in the directory.
+     * Currently deleting only from local directory as files from remote should not be deleted as that is taken care by garbage collection logic of remote directory
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        logger.trace("deleteFile() called {}", name);
+        if (isTempFile(name)) {
+            localDirectory.deleteFile(name);
+        } else {
+            /*
+            Not deleting from localDirectory directly since it causes a race condition when the localDirectory deletes a file, and it ends up in pendingDeletion state.
+            Meanwhile, fileCache on removal deletes the file directly via the Files class and later when the directory tries to delete the files pending for deletion (which happens before creating a new file), it causes NoSuchFileException and new file creation fails
+            */
+            fileCache.remove(localDirectory.getDirectory().resolve(name));
+        }
+    }
+
+    /**
+     * Returns the byte length of a file in the directory.
+     * Throws {@link NoSuchFileException} or {@link FileNotFoundException} in case file is not present locally and in remote as well
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public long fileLength(String name) throws IOException {
+        logger.trace("fileLength() called {}", name);
+        long fileLength;
+        Path key = localDirectory.getDirectory().resolve(name);
```

Comment:
is there a case where key is null?

Reply:
Key won't be null if local directory is not null. We already have checks for non null values of local directory in the Composite Directory constructor